includes/Core/Util/Entity_Factory.php

```diff
@@ -60,8 +60,54 @@ public static function from_context() {
 
 		// Otherwise, run frontend-specific `WP_Query` logic.
 		if ( $wp_the_query instanceof WP_Query ) {
-			return self::from_wp_query( $wp_the_query );
+			$entity = self::from_wp_query( $wp_the_query );
+
+			if ( null === $entity ) {
+				return null;
+			}
+
+			if ( ! defined( 'AMP__VERSION' ) ) {
+				return $entity;
+			}
+
+			$url_parts    = wp_parse_url( $_SERVER['REQUEST_URI'] ); // phpcs:ignore
+			$new_url_tail = '';
+
+			// check if the $url has amp query param.
+			if ( strpos( $url_parts['query'], 'amp' ) !== false ) {
+				$new_url_tail = '?amp=1';
+			}
```

Comment:
Checking the query argument using `strpos` is not sufficient because it can produce false positives when `amp` sequence is present in a query arg value (`?product_type=lamp`) or is a part of another query key (`?lamp=true`). Let's use the [`wp_parse_str`](https://developer.wordpress.org/reference/functions/wp_parse_str/) function to parse query arguments and check whether the `amp` key exists.